Connor Goosen<br>
GSNCON001

# Analytical Problems:

## 1.a
Given the formula: </br>
$$x_{n+1} = x_n - (\cos{x_n})(\sin{x_n}) + R\cos^2{x_n}$$
knowing that the formula was derived by means of the Newton-Raphson method:
$$x_{n+1} = x_n - \frac{f(x_n)}{f^\prime(x_n)}$$

allows us to say that:
$$\frac{f(x_n)}{f^\prime(x_n)} = (\cos{x_n})(\sin{x_n}) + R\cos^2{x_n}$$

Additionally, from calculus we know that $\int{\frac{f^\prime(x)}{f(x)}} = \ln|f(x)|$, therefore by inverting both sides of the equation gives us:
$$\frac{f^\prime(x)}{f(x)} = \frac{1}{\cos{x_n})(\sin{x_n}) + R\cos^2{x_n}}$$
Therefore, integrating both sides:
$$\int{\frac{f^\prime(x)}{f(x)}} = \int{\frac{1}{\cos{x_n})(\sin{x_n}) + R\cos^2{x_n}}}$$
<br>
$$\ln{|f(x)|} = \ln{|\tan(x_n)-R|}$$
<br>
$$\Rightarrow f(x_n) = \tan(x_n)-R$$

## 1.b
This formula can be used to find the root of $tan(x_n)-R$, the original function.

# Numerical Problems:
The task is to find the root to the function: $f(x) = x^3-10$ using multiple approximation algorithms and compare their convergence to the root.

In [56]:
function f(x) #function
    return x^3 -10
end

function ff(x) #first derivative of function
    return 3x^2
end

function fff(x) #second derivative of function
    return 6x
    end;

In [62]:
function Halley(xn)
   xn = xn - (2.0*f(xn)*ff(xn))/((2*ff(xn)^2) - f(xn)*fff(xn)) 
end

function Newton(xn)
    xn=xn-(f(xn)/ff(xn))
end

function Bisection(xl, xr)
    xm = (xl+xr)/2.0
    tmp = 0
    approx = []
    while xm != tmp
        tmp = xm
        (f(xm)>0) ? xr = xm : xl=xm
        append!(approx, xm)
        xm = (xr+xl)/2
    end
    return approx
    end;

In [63]:
function rootfinder(any_func::Function)
x0 = 2.0
tmp = 0.0
approx = [2.0]
while x0 != tmp
    tmp = x0
    x0 = any_func(x0)
    append!(approx, x0)
end
approx
end

halley = rootfinder(Halley)
newton=rootfinder(Newton);
bisection = Bisection(2.0,4.0);

## 2a
Implementing the Halley Method as seen in the ```function Halley(xn)``` method above
yeilds:

In [64]:
halley

5-element Array{Float64,1}:
 2.0               
 2.1538461538461537
 2.1544346900025926
 2.154434690031884 
 2.154434690031884 

Which can be seen to converge after just four iterations and gives the final approximaiton of the root as $x_n = 2.154434690031884$

## 2b
Now implementing the Newton-Raphson method on the same funciton gives the approximations:

In [67]:
newton

6-element Array{Float64,1}:
 2.0               
 2.1666666666666665
 2.1545036160420774
 2.1544346922369133
 2.154434690031884 
 2.154434690031884 

Which takes five iterations to converge to the same value, showing it to be a slightly slower method.

## 2c
Finally comparing Halley's method to the Bisection algorithm taught in class:
We see that by implementing the algorithm:

In [68]:
bisection

53-element Array{Any,1}:
 3.0               
 2.5               
 2.25              
 2.125             
 2.1875            
 2.15625           
 2.140625          
 2.1484375         
 2.15234375        
 2.154296875       
 2.1552734375      
 2.15478515625     
 2.154541015625    
 ⋮                 
 2.1544346900323035
 2.154434690032076 
 2.1544346900319624
 2.1544346900319056
 2.154434690031877 
 2.1544346900318914
 2.1544346900318843
 2.1544346900318807
 2.1544346900318825
 2.1544346900318834
 2.154434690031884 
 2.1544346900318834

That we get an array of length 53 which means that it takes the algorithm 52 iterations in order to converge. When it does converge however, it can be noticed that it still oscillates between $2.154434690031884$, the result given by both the Halley and Newton methods and $2.1544346900318834$ showing it to be a less accurate and much slower algorithm to the Halley or Newton methods.

## Summary of numerical problem:
From this simple analysis it is clear to see that the Halley method of root finding is the fastest of the three and is more acccurate than the bisection method.

| Method: | Halley's Method | Newton's Method | Bisection Method |
| ------- | --------------- | --------------- | ---------------- |
| Final result:| 2.154434690031884 | 2.154434690031884 | 2.1544346900318834 |
| Number of iterations required: | 4 | 5 | 52 |